server/src/dedup_requests.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022-2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::{collections::HashMap, hash::Hash, sync::RwLock};
+
+use tokio::sync::mpsc::Sender;
+
+type Notifier<T> = Sender<T>;
+
+#[derive(Debug)]
+struct Notifiers<T> {
+    notifiers: RwLock<Vec<Notifier<T>>>,
+}
+
+impl<T> Notifiers<T> {
+    pub fn new(notifier: Notifier<T>) -> Self {
+        let notifiers = vec![notifier];
+        Self {
+            notifiers: RwLock::new(notifiers),
+        }
+    }
+
+    pub fn add_notifier(&self, notifier: Notifier<T>) {
+        self.notifiers.write().unwrap().push(notifier);
+    }
+}
+
+#[derive(Debug)]
+pub struct RequestNotifiers<K, T>
+where
+    K: PartialEq + Eq + Hash,
+{
+    inner: RwLock<HashMap<K, Notifiers<T>>>,
+}
+
+impl<K, T> Default for RequestNotifiers<K, T>
+where
+    K: PartialEq + Eq + Hash,
+{
+    fn default() -> Self {
+        Self {
+            inner: RwLock::new(HashMap::new()),
+        }
+    }
+}
+
+impl<K, T> RequestNotifiers<K, T>
+where
+    K: PartialEq + Eq + Hash,
+{
+    /// Insert a notifier for the given key.
+    pub fn insert_notifier(&self, key: K, notifier: Notifier<T>) -> RequestResult {
+        // First try to read the notifiers, if the key exists, add the notifier to the
+        // notifiers.
+        let notifiers = self.inner.read().unwrap();
+        if notifiers.contains_key(&key) {
+            notifiers.get(&key).unwrap().add_notifier(notifier);
```

Comment:
The check on L55 may be wrong when L56 is reached. Maybe we should a method called insert_if_key_exists for notifiers.